### DISCUSSION TITLE
SwiftUI + MVVM ?

### DISCUSSION THREAD

**Discussion by 00yhsp** [DeveloperAcademy-POSTECH/.github#16] [Q&A]
논쟁이 많은 주제인 걸로 알고 있는데
마침 NC1 구현 중 CoreData 관련 코드들을 뷰모델로 이관하며 뷰 단일 구조에 비해 오히려 코드가 더러워지는 경험을 해서
여러분들의 경험담, 생각이 궁금해 여쭤봅니다.

**Comment by insub4067** [comment]
안녕하세요 지나가던 1기 입니다. 저도 아키택쳐에 관심이 많은 편인데요. 뷰모델을 사용했을 때 어떤 이유로 단일 뷰 구조에 비해 비대해졌는지 이유가 궁금합니다.

**Reply by 00yhsp** [reply]
@FetchRequest 프로퍼티 래퍼를 이용해 처리할 수 있었던 CoreData - View 간 상태 관리 로직을 뷰모델로 이관해 작성하면서 뷰 단일 구조에 비해 코드 양이 비대해지고 가독성도 떨어지는 경험을 했습니다. 아직 Swift에 익숙치 않아 코드 퀄리티의 문제일 수 있겠지만,, 상태 관리를 지원하는 SwiftUI에서 뷰모델의 필요성이 있다면 무엇인가에 대해 의문이 생겨 질문하게 되었습니다!

**Comment by hhajime** [comment]
많은 [논쟁](https://www.reddit.com/r/SwiftUI/comments/194wiun/why_should_i_use_mvvm/)이 여전히 있지만, 프로젝트 규모에 따라 나뉠 것 같아요! 

뷰가 수십개 이상이고 프로젝트가 무거워 질 수록 테스트 용이성, 모듈화, 코드 중복 방지 등을 위해서 뷰와 로직을 분리하는 것이 좋아보이고, 간단한 앱의 경우 MV와 같은 간단한 패턴을 적용해도 상관없다라는 의견이 많아 보이네요.

이 외에도 여러 아키텍쳐가 있지만 ..  제 사견으로는 해당 프로젝트의 특성, 구성원의 수준, 비즈니스 영향도에 따라 아키텍쳐의 방향성이 달라지고
프로젝트의 규모가 커지면 그에 따라 필요성이 느껴질 것이고, 자연스럽게 변화시키는 것이 좋아 보입니다..!

처음부터 호랑이를 잡지 않듯이요!

MVVM에 대한 중요성은 해당 영상 참고하셔도 좋을 것 같습니다. 
한국어 자막이 포함된 [Stanford CS193p 강의](https://www.youtube.com/watch?v=WZj86AE9QxM&ab_channel=%EA%B7%93%EC%86%8D%EB%A7%90%EC%9D%98%EC%9E%90%EB%A7%89%EB%82%98%EB%9D%BC)입니다!

**Reply by hhajime** [reply]
https://gist.github.com/unnnyong/439555659aa04bbbf78b2fcae9de7661

예전에 @unnnyong 이 번역해주신 아티클인데, 저도 MVVM에서 SwiftUI의 최신 프로퍼티 래퍼들을 활용하기가 힘들었던 경험이 있어요!
같이 참고하셔도 좋을 것 같네요 :)
